### PR TITLE
Migrate SPDYRoundTripper to use the standard client-go.Transport instead of custom dialer

### PR DESCRIPTION
**Pull request by mikedanese**
These simplifications greatly reduces the complexity of implementing  [KEP-2718: Client Executable Proxy](https://github.com/kubernetes/enhancements/tree/e96e05d97da30e22502fa563a76e2e0a4593546b/keps/sig-auth/2718-20210511-client-exec-proxy) by removing custom dialing of upgrade backends.

First two commits under review in: https://github.com/kubernetes/kubernetes/pull/113637

## Commit: support upgrade requests on cached transports

The stdlib transport kind of does this:

https://cs.opensource.google/go/go/+/master:src/net/http/transport.go;l=840;drc=3615eb16af57b832a3b6d957d86b84b656865a14

But it's not as aggressive as we need it to be because it only forces H1 on websocket upgrades. We need to force H1 on all upgrades. Since it's a connection error to use the "Upgrade" header in an http2 request (https://datatracker.ietf.org/doc/html/rfc7540#section-8.1.2.2), we use H1 for all requests with an upgrade header. 

## Commit: use a standard transport to establish spdy streams

From http://pkg.go.dev/net/http#Request:

> As of Go 1.12, the Body will also implement io.Writer on a successful "101 Switching Protocols" response, as used by WebSockets and HTTP/2's "h2c" mode.

I assume the reason we didn't use net/http client and instead wrote custom dialing logic for the SPDY round tripper was because this code was implemented before go 1.12.

We should just use the Transport now. We put a ton of effort into configuring it in k8s.io/client-go and we always have to port that effort to the upgrade dialers if we maintain this fork.

/kind cleanup
/sig api-machinery

```release-note
NONE
```

### CODE REVIEW

**Comment by aojea** [comment]
```
I1105 00:14:38.183] Summarizing 1 Failure:
I1105 00:14:38.183]   [PANICKED!] [sig-cli] Kubectl client Simple pod [It] should support exec through an HTTP proxy
I1105 00:14:38.183]   /usr/local/go/src/runtime/panic.go:260
```

sounds related :/


I have confess this is really a scary thing 😄

**Comment by liggitt** [comment]
The unit test failures look very related. I won't be able to get to this until after code freeze

File: staging/src/k8s.io/client-go/transport/cache.go [hunk 1]
```diff
@@ -138,6 +147,21 @@ func (c *tlsTransportCache) get(config *Config) (http.RoundTripper, error) {
...
+func (tr *upgradeAwareTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	if v := req.Header.Get("Upgrade"); v != "" {
+		return tr.h1.RoundTrip(req)
+	}
+	return tr.h2.RoundTrip(req)
+}
+
+func (tr *upgradeAwareTransport) WrappedRoundTripper() http.RoundTripper {
+	return tr.h2
+}
```
Comment (enj) [inline staging/src/k8s.io/client-go/transport/cache.go:163]: This does not look correct to me (at least not in all contexts).  Looking through the usages of `WrappedRoundTripper`, we use it in a few places, one of them being request cancellation.  In such cases, the correct round tipper to return is based on the request.

File: staging/src/k8s.io/client-go/transport/cache.go [hunk 2]
```diff
@@ -121,14 +121,23 @@ func (c *tlsTransportCache) get(config *Config) (http.RoundTripper, error) {
...
 		TLSClientConfig:     tlsConfig,
 		MaxIdleConnsPerHost: idleConnsPerHost,
 		DialContext:         dial,
 		DisableCompression:  config.DisableCompression,
-	})
+	}
+	h2tr := h1tr.Clone()
+
+	utilnet.SetOldTransportDefaults(h1tr)
+	utilnet.SetTransportDefaults(h2tr)
```
Comment (enj) [inline staging/src/k8s.io/client-go/transport/cache.go:135]: Add go docs explaining what you are doing here.

File: staging/src/k8s.io/client-go/transport/cache.go [hunk 3]
```diff
@@ -138,6 +147,21 @@ func (c *tlsTransportCache) get(config *Config) (http.RoundTripper, error) {
 	return transport, nil
 }
 
+type upgradeAwareTransport struct {
+	h1, h2 http.RoundTripper
```
Comment (enj) [inline staging/src/k8s.io/client-go/transport/cache.go:151]: Use more explicit field names with go docs.

File: staging/src/k8s.io/client-go/transport/cache.go [hunk 4]
```diff
@@ -138,6 +147,21 @@ func (c *tlsTransportCache) get(config *Config) (http.RoundTripper, error) {
...
+}
+
+func (tr *upgradeAwareTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	if v := req.Header.Get("Upgrade"); v != "" {
```
Comment (enj) [inline staging/src/k8s.io/client-go/transport/cache.go:155]: I think it would be more correct to check the map for the presence of the key instead of silently skipping this `if` statement if the upgrade header is set to the empty string.

File: pkg/kubelet/server/server_test.go [hunk 5]
```diff
@@ -1112,9 +1107,6 @@ func testExecAttach(t *testing.T, verb string) {
 			require.NoError(t, err, "POSTing")
 			defer resp.Body.Close()
 
-			_, err = io.ReadAll(resp.Body)
-			assert.NoError(t, err, "reading response body")
-
```
Comment (enj) [inline pkg/kubelet/server/server_test.go:1117]: Why was this removed?

File: staging/src/k8s.io/client-go/transport/spdy/spdy.go [hunk 6]
```diff
@@ -94,7 +86,6 @@ func Negotiate(upgrader Upgrader, client *http.Client, req *http.Request, protoc
 	if err != nil {
 		return nil, "", fmt.Errorf("error sending request: %v", err)
 	}
-	defer resp.Body.Close()
```
Comment (enj) [inline staging/src/k8s.io/client-go/transport/spdy/spdy.go:97]: Why was this removed?

File: staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go [hunk 7]
```diff
@@ -17,286 +17,44 @@ limitations under the License.
...
-		Proxier: proxier,
-	})
-}
-
-// NewRoundTripperWithConfig creates a new SpdyRoundTripper with the specified
-// configuration.
-func NewRoundTripperWithConfig(cfg RoundTripperConfig) *SpdyRoundTripper {
-	if cfg.Proxier == nil {
-		cfg.Proxier = utilnet.NewProxierWithNoProxyCIDR(http.ProxyFromEnvironment)
+func NewRoundTripper(rt http.RoundTripper) *SpdyRoundTripper {
```
Comment (enj) [inline staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:47]: Trying to think if we need timeout handling:

https://github.com/kubernetes/kubernetes/blob/ac22aff150e7ae6eaabe0c63f3e27b1924a0b790/staging/src/k8s.io/client-go/rest/transport.go#L37-L45

File: staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go [hunk 8]
```diff
@@ -17,286 +17,44 @@ limitations under the License.
...
-	pingPeriod time.Duration
+	rt http.RoundTripper
 }
 
-var _ utilnet.TLSClientConfigHolder = &SpdyRoundTripper{}
 var _ httpstream.UpgradeRoundTripper = &SpdyRoundTripper{}
-var _ utilnet.Dialer = &SpdyRoundTripper{}
 
 // NewRoundTripper creates a new SpdyRoundTripper that will use the specified
 // tlsConfig.
```
Comment (enj) [inline staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:46]:
```suggestion
```

File: staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go [hunk 9]
```diff
@@ -17,286 +17,44 @@ limitations under the License.
...
-		return ""
-	}
-	credentials := proxyURL.User.String()
-	encodedAuth := base64.StdEncoding.EncodeToString([]byte(credentials))
-	return fmt.Sprintf("Basic %s", encodedAuth)
+func (s *SpdyRoundTripper) WrappedRoundTripper() http.RoundTripper {
+	return s.rt
 }
 
 // RoundTrip executes the Request and upgrades it. After a successful upgrade,
```
Comment (enj) [inline staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:60]: This comment looks wrong, is it talking about `NewConnection`?

File: staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go [hunk 10]
```diff
@@ -350,7 +93,12 @@ func (s *SpdyRoundTripper) NewConnection(resp *http.Response) (httpstream.Connec
 		return nil, fmt.Errorf("unable to upgrade connection: %s", responseError)
 	}
 
-	return NewClientConnectionWithPings(s.conn, s.pingPeriod)
+	rwc, ok := resp.Body.(io.ReadWriteCloser)
+	if !ok {
+		return nil, fmt.Errorf("internal error: 101 switching protocols response with non-writable body")
+	}
+
+	return NewClientConnectionWithPings(&conn{rwc: rwc}, 5*time.Second)
```
Comment (enj) [inline staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:101]: The `net.Conn` in `conn` is `nil` here.  Is that intentional?  This code used to use `s.conn` which used to be set in `RoundTrip`.

**Review by enj — COMMENTED** [review]
Do we have enough integration tests around SPDY usage to feel confident about such a change?

File: staging/src/k8s.io/client-go/transport/cache.go (hunk 4 above)
Comment (enj) [reply staging/src/k8s.io/client-go/transport/cache.go:155]: `httpstream.IsUpgradeRequest` might be appropriate?

**Comment by enj** [comment]
Related, I think we do a similar refactor to our server side code.  I believe this upgrade logic can be handled by the std lib reverse proxy if you use distinct transports for http1 and http2 (I used that approach is a different project):

https://github.com/kubernetes/kubernetes/blob/62e9df085a82a5e49c435e5b66f988095a52fe3b/staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go#L308-L321

**Comment by liggitt** [comment]
we also need to be really careful about making client changes that require server changes, or vice versa. our tests need to not only cover current clients → current servers, but ensure old clients are not broken, and that new clients can still talk to old servers

File: staging/src/k8s.io/client-go/transport/cache.go (hunk 1 above)
Comment (mikedanese) [reply staging/src/k8s.io/client-go/transport/cache.go]: Pushed different implementation of this. PTAL.

File: staging/src/k8s.io/client-go/transport/cache.go (hunk 3 above)
Comment (mikedanese) [reply staging/src/k8s.io/client-go/transport/cache.go]: Obsolete.

File: pkg/kubelet/server/server_test.go (hunk 5 above)
Comment (mikedanese) [reply pkg/kubelet/server/server_test.go:1117]: Today, we create this response "manually" with http.ReadResponse here: https://github.com/kubernetes/kubernetes/blob/f87231003aeb43bf884cf9dc10b1247b8ae5cbb8/staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go#L317

The body is http.NoBody because the status is 101 switching protocols: https://cs.opensource.google/go/go/+/master:src/net/http/transfer.go;l=561;drc=3a41094107237535673bdbbdaabc10e54cb2b67c

In this PR, the response is now created by http.Transport. For 101 switching protocols, the transport sets up the body after calling http.ReadResponse: https://cs.opensource.google/go/go/+/master:src/net/http/transport.go;l=2319-2321;drc=3a41094107237535673bdbbdaabc10e54cb2b67c

So now, resp.Body is something that we want to leave open while we're streaming.

File: staging/src/k8s.io/client-go/transport/spdy/spdy.go (hunk 6 above)
Comment (mikedanese) [reply staging/src/k8s.io/client-go/transport/spdy/spdy.go:97]: Same as above. Now, resp.Body is something that we want to leave open while we're streaming.

File: staging/src/k8s.io/client-go/transport/cache.go (hunk 4 above)
Comment (mikedanese) [reply staging/src/k8s.io/client-go/transport/cache.go]: PTAL at new upgrade check.

File: staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go (hunk 9 above)
Comment (mikedanese) [reply staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:62]: Fixed.

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR

File: test/e2e/kubectl/kubectl.go [hunk 11]
```diff
@@ -472,8 +470,12 @@ var _ = SIGDescribe("Kubectl client", func() {
 				framework.Failf("--host variable must be set to the full URI to the api server on e2e run.")
 			}
 
-			ginkgo.By("Starting goproxy")
-			testSrv, proxyLogs := startLocalProxy()
+			ginkgo.By("Starting http_proxy")
+			var proxyLogs bytes.Buffer
+			testSrv := httptest.NewServer(utilnet.NewHTTPProxyHandler(func(req *http.Request) bool {
```
Comment (aojea) [inline test/e2e/kubectl/kubectl.go:475]: do we need this new library here or shouldn't be enough with `NewSingleHostReverseProxy` only?

File: staging/src/k8s.io/apimachinery/pkg/util/net/http.go [hunk 12]
```diff
@@ -697,3 +699,109 @@ func makeQuotedString(s string) string {
...
+	}
+
+	b, err := httputil.DumpRequest(req, false)
+	if err != nil {
+		klog.ErrorS(err, "Failed to dump request", "host", req.Host)
+	} else {
+		klog.Infof("Proxy Request: %s", string(b))
+	}
+
+	if req.Method != http.MethodConnect {
```
Comment (aojea) [inline staging/src/k8s.io/apimachinery/pkg/util/net/http.go:781]: is this not handled now internally by the std lib?

File: staging/src/k8s.io/apimachinery/pkg/util/net/http.go [hunk 13]
```diff
@@ -173,11 +176,28 @@ func pingTimeoutSeconds() int {
...
+	t2, err := http2.ConfigureTransports(tInner)
 	if err != nil {
-		return err
+		return nil, err
 	}
+
+	// Because the stdlib transport tries to send upgrades over HTTP/2, we wrap
+	// it in our own h2RoundTripper which returns http.ErrSkipAltProtocol on
+	// upgrade requests. This causes the tOuter transport to handle the request
+	// (which only does HTTP/1.1_
```
Comment (aojea) [inline staging/src/k8s.io/apimachinery/pkg/util/net/http.go:199]: this is the part that I'm more worried about, suddenly the same client will handle 2 transports, one for upgraded requests and other for the rest, ...  I have not clear the consequences of this
why do we need to automate this? why not fix the spdy rountripper only and let the users configured it independently?

File: staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go (hunk 7 above)
Comment (aojea) [reply staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:49]: I don't think we need it, I think this should be used

```
 tr, err := rest.TransportFor(config)
if err != nil {
  //
}
 spdytr := NewRoundtripper(tr)
...
```

File: staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go [hunk 14]
```diff
@@ -350,7 +95,12 @@ func (s *SpdyRoundTripper) NewConnection(resp *http.Response) (httpstream.Connec
 		return nil, fmt.Errorf("unable to upgrade connection: %s", responseError)
 	}
 
-	return NewClientConnectionWithPings(s.conn, s.pingPeriod)
+	rwc, ok := resp.Body.(io.ReadWriteCloser)
```
Comment (aojea) [inline staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:98]: can we add a comment explaining that the Upgrade request was successful at this point and the response body should be bidirectional?

File: staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go (hunk 10 above)
Comment (aojea) [reply staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:103]: I have created a handy library for this https://github.com/aojea/rwconn , is the one used in the kcp project to tunnel exec, logs, ... https://github.com/kcp-dev/kcp/pull/873

**Comment by enj** [comment]
> But it's not as aggressive as we need it to be because it only forces H1 on websocket upgrades. We need to force H1 on all upgrades. Since it's a connection error to use the "Upgrade" header in an http2 request (https://datatracker.ietf.org/doc/html/rfc7540#section-8.1.2.2), we use H1 for all requests with an upgrade header.

Can we just fix the standard library to do this on all upgrades as well?

File: staging/src/k8s.io/client-go/transport/spdy/spdy.go (hunk 6 above)
Comment (enj) [reply staging/src/k8s.io/client-go/transport/spdy/spdy.go:97]: Do we need to close the response body in error cases?

File: staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go (hunk 10 above)
Comment (enj) [reply staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:103]: (this is outstanding)

**Comment by aojea** [comment]
> FWIW I have had success in the past with using split transports for upgrade and non-upgrade paths. I am generally in favor of using more standard lib code over our own custom stuff.

interesting, any links to learn more?

I still don't see the use case/requirement to have the duality embedded in the client, users can create one client for each, but I may have missed it, can you please remind me why is it needed?

**Comment by enj** [comment]
> interesting, any links to learn more?

I mentioned it briefly in https://github.com/kubernetes/kubernetes/pull/113639#issuecomment-1305737068, but compare the upgrade code path used by the aggregator proxy (a bug in `tryUpgrade` led to #71411):

https://github.com/kubernetes/kubernetes/blob/3a8e2e399b2bb8786d86fbfcbe6434a9774dcd08/staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go#L213-L220

vs. [this impersonation proxy](https://github.com/vmware-tanzu/pinniped/blob/f4c9202f4914db0141e4fe6e35b27e30defeb0d1/internal/concierge/impersonator/impersonator.go#L528-L534).

Both of these examples involve a reverse proxy placed in front of an API server.  The proxy handles auth and then sets headers before sending the request to the API server.  The proxy has to work with upgrade requests.  The former has a bunch of custom code for this, the latter simply delegates to the std lib by constructing two transports (one that is http1 only and the other that is http2 only).
 
> I still don't see the use case/requirement to have the duality embedded in the client, users can create one client for each, but I may have missed it, can you please remind me why is it needed?

In the end I think you will have the same logic, just in different places.  There might be a better place to put this logic but with the current approach unwrapping transports works correctly, which was one of the bigger concerns I had.

**Comment by aojea** [comment]
> There might be a better place to put this logic but with the current approach unwrapping transports works correctly, which was one of the bigger concerns I had.

This will be much harder to support
